### PR TITLE
Tmz 1.6.0 navajo

### PR DESCRIPTION
**Pull request by t-m-z**
encoding and decoding using the Navajo Code Talkers' Dictionary REVISED 15 JUNE 1945 (DECLASSIFIED UNDER DEPARTMENT OF DEFENSE DIRECTIVE 5200.9)
from 
https://www.history.navy.mil/research/library/online-reading-room/title-list-alphabetically/n/navajo-code-talker-dictionary.html
https://www.ancestrycdn.com/aa-k12/1112/assets/Navajo-Code-Talkers-dictionary.pdf

tests checked with 
https://kryptografie.de/kryptografie/chiffre/navajo.htm
https://www.dcode.fr/navajo-code

hint
1. the alphabet-table at kryptografie.de has errors - misspelling
2. the whole dictionary is not implemented - only the alphabet
3. kryptografie uses a short version of the alphabet
4. dcode uses the alphabet table from the dictionary and encodes letters randomly

### CODE REVIEW

File: lib/logic/tools/crypto_and_encodings/navajo.dart [hunk 1]
```diff
@@ -0,0 +1,823 @@
...
+Map NAVAJO_DECODE_DICTIONARY = switchMapKeyValue(NAVAJO_ENCODE_DICTIONARY);
+
+Map NAVAJO_DECODE_ALPHABET = switchMapKeyValue(NAVAJO_ENCODE_ALPHABET);
+
+String shrinkText(String input){
+  return input
+      .replaceAll('COMMANDING GEN.', 'COMMANDINGGEN')
+      .replaceAll('MAJOR GEN.', 'MAJORGEN')
+      .replaceAll('BRIGADIER GEN.', 'BRIGADIERGEN')
+      .replaceAll('LT. COLONEL', 'LTCOLONEL')
```
Comment (S-Man42) [inline lib/logic/tools/crypto_and_encodings/navajo.dart:665]: IIN line 51 you write LT.COLONEL with a dot... Problem?

Comment (t-m-z) [reply lib/logic/tools/crypto_and_encodings/navajo.dart:665]: checked and changed

**Comment by S-Man42** [comment]
Several issues here:

1. It's not clear to what "only alphabet" does. Can you make an example please? "Only Alphabet" only returns the input....
grafik

2. Click on second Switch directly after opening the tool: Both switches move, not only the second. They only become independent when you switch the first one first
3. When typing an input, and then switching the source removes the input... Seems not very comfortable...
4. Some strange misbehaviour occurs when play long enough with the switches. In this example I guess it's the output from encryption not from decryption:
grafik

I guess your state handling is not proper.

**Comment by S-Man42** [comment]
Ah, I believe I found the problem: Your second switch sets _currentMode not _currentSource

**Comment by t-m-z** [comment]
1. Behaviour of 2nd switch fixed
2. change "Source" to "Using" - this should make it better to understand
3. Switching between encrypt and decrypt does also switch to a new controller. Hence the input seems to be deleted. But this behaviour is found in several tools

PR done

File: lib/logic/tools/crypto_and_encodings/navajo.dart [hunk 2]
```diff
@@ -0,0 +1,823 @@
...
+  NAVAJO_DECODE_ALPHABET['SHI-DA'] = 'U';
+  NAVAJO_DECODE_ALPHABET['BSEH-DO-GLIZ'] = 'Z';
+
+  List<String> result = new List<String>();
+  if (cipherText == null || cipherText == '') return '';
+
+  cipherText.toUpperCase().split(' ').forEach((element) {
+    if (NAVAJO_DECODE_ALPHABET[element] == null)
+      if (useOnlyAlphabet)
+        result.add('?');
```
Comment (S-Man42) [inline lib/logic/tools/crypto_and_encodings/navajo.dart:785]: Here two options:

1. Normally we ignore input which do not deliver a result. For example "AAA 111 222" simply ignores the "AAA" block in Vanity and does not return "unknown". The idea behind is to enter any text and return only the relevant information for direct copy without the need to remove some "?" chars afterwards manually. Think about a listing which is: "And navajo indian said: A-WOH SHI-DA, but I didn't understand him". In that case you could input the entire text and should only return "TU". I know, there are tools that behave differently, like the Alphabet Values which signalize that "Ä" is not part of the chosen Standard Alphabet or something. But in your tool I would tend to the first way
2. However, if you decide for the second way, please use the common constant "UNKNOWN_ELEMENT" instead.

File: lib/logic/tools/crypto_and_encodings/navajo.dart [hunk 3]
```diff
@@ -0,0 +1,823 @@
+// https://www.history.navy.mil/research/library/online-reading-room/title-list-alphabetically/n/navajo-code-talker-dictionary.html
+// https://www.ancestrycdn.com/aa-k12/1112/assets/Navajo-Code-Talkers-dictionary.pdf
+// https://kryptografie.de/kryptografie/chiffre/navajo.htm
```
Comment (S-Man42) [inline lib/logic/tools/crypto_and_encodings/navajo.dart:3]: please DO NOT use kryptografie.de as source!

**Comment by S-Man42** [comment]
Input Encrypt:
"sdf sdf" -> "DIBEH BE MA-E  DIBEH BE MA-E"

Taking this output as input Decrypt:
"DIBEH BE MA-E  DIBEH BE MA-E" -> "SDF?SDF". 

The space becomes a "?".

The encrypted output now has two spaces. I guess you make something like "inputtext".split(' '). In that case you create and empty object (between the two spaces). Better sanitize your input. Two options:
1.` input.replaceAll('\s+', ' ')` // reduces all whitespaces, no matter how long to one single space before processing
2. `input.split(' ').where((element) => element != null && !element.isEmpty())...` // removes the empty elements from split result

A third (maybe additional) option can be to ignore the space in encrypt output. Is it really necessary to return a double space? ;)

File: lib/logic/tools/crypto_and_encodings/navajo.dart (hunk 3 above)
Comment (t-m-z) [reply lib/logic/tools/crypto_and_encodings/navajo.dart:3]: done

File: lib/widgets/searchstrings/searchstrings_common.dart [hunk 4]
```diff
@@ -205,6 +205,7 @@ const SEARCHSTRING_COMMON_MAYANUMBERS = 'mayas majas vigesimalsystem 20';
 const SEARCHSTRING_COMMON_MEXICANARMYCIPHERWHEEL = '  ';
 const SEARCHSTRING_COMMON_MORSE = 'samuel morse ';
 const SEARCHSTRING_COMMON_MULTIDECODER = '';
+const SEARCHSTRING_COMMON_NAVAJO = 'navajo';
```
Comment (S-Man42) [inline lib/widgets/searchstrings/searchstrings_common.dart:208]: i'd add "najaho" as well. Saw that sometimes (for example my old climbing gear is called "Petzl Navaho" - they call their gear after indian tribes)

Comment (t-m-z) [reply lib/widgets/searchstrings/searchstrings_common.dart:208]: check - done

**Comment by t-m-z** [comment]
> DIBEH BE MA-E DIBEH BE MA-E

Good point - the idea behind: how to deal with words encoded letterwise? The double whitespace should symbolize, there starts the next word.
One option could be to simply the process and ignore these spaces.
One option could be to make it more sophisticated.

I will think about.

File: lib/logic/tools/crypto_and_encodings/navajo.dart (hunk 2 above)
Comment (t-m-z) [reply lib/logic/tools/crypto_and_encodings/navajo.dart:785]: done

**Comment by S-Man42** [comment]
KISS: Keep it simple and stupid :)

**Comment by S-Man42** [comment]
Decision here? Can this be merged or is it WIP?

**Comment by t-m-z** [comment]
can be merged

Sent from MailDroid

-----Original Message-----

**Pull request merged**